experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,699 @@
+export SpecOpen, ambient, gens, complement, npatches, affine_patches, intersections, name, intersect, issubset, closure, find_non_zero_divisor, is_non_zero_divisor, is_dense
+
+export StructureSheafRing, scheme, domain, OO
+
+export StructureSheafElem, domain, restrictions, patches, restrict, npatches
+
+export SpecOpenMor, maps_on_patches, restriction, identity_map, preimage, generic_fractions, pullback, maximal_extension
+
+@Markdown.doc """
+    SpecOpen{BRT, BRET, RT, RET, MST} <: Scheme{BRT, BRET}
+
+Zariski open subset ``U`` of an affine scheme ``X = Spec(R)``. 
+This stores a list of generators ``f₁,…,fᵣ`` of an ideal 
+``I`` defining the complement ``Z = X ∖ U``. 
+The scheme ``X`` is referred to as the *ambient scheme* and 
+the list ``f₁,…,fᵣ`` as the *generators* for ``U``.
+"""
+@attributes mutable struct SpecOpen{BRT, BRET, RT, RET, MST} <: Scheme{BRT, BRET}
+  X::Spec{BRT, BRET, RT, RET, MST} # the ambient scheme
+  gens::Vector{RET} # a list of functions defining the complement of the open subset
+
+  # fields used for caching
+  name::String
+  patches::Vector{Spec{BRT, BRET, RT, RET, MST}}
+  intersections::Dict{Tuple{Int}, Spec{BRT, BRET, RT, RET, MST}}
+  complement::Spec{BRT, BRET, RT, RET, MST}
+
+  function SpecOpen(X::Spec{BRT, BRET, RT, RET, MST}, f::Vector{RET}; name::String="") where {BRT, BRET, RT, RET, MST}
+    for a in f
+      parent(a) == base_ring(OO(X)) || error("element does not belong to the correct ring")
+    end
+    if length(name) > 0 
+      U = new{BRT, BRET, RT, RET, MST}(X, f)
+      set_name!(U, name)
+      return U
+    end
+    return new{BRT, BRET, RT, RET, MST}(X, f)
+  end
+end
+
+@Markdown.doc """
+    ambient(U::SpecOpen)
+
+Return the ambient scheme ``X`` of a Zariski open subset ``U ⊂ X``.
+"""
+ambient(U::SpecOpen) = U.X
+
+@Markdown.doc """
+    npatches(U::SpecOpen)
+
+Return the number of generators stored for describing the complement of ``U``.
+"""
+npatches(U::SpecOpen) = length(U.gens)
+
+@Markdown.doc """
+    gens(U::SpecOpen) = U.gens
```

Comment:
```suggestion
    gens(U::SpecOpen)
```
Don't put implementation details into docstrings, and least of all in that spot :-)